tyk_reverse_proxy_clone.go

```diff
@@ -37,30 +37,30 @@ func GetURLFromService(spec *APISpec) (*apidef.HostList, error) {
 		sd := ServiceDiscovery{}
 		sd.New(&spec.Proxy.ServiceDiscovery)
 		data, err := sd.GetTarget(spec.Proxy.ServiceDiscovery.QueryEndpoint)
-		if err == nil {
-			// Set the cached value
-			if data.Len() == 0 {
-				spec.HasRun = true
-				spec.ServiceRefreshInProgress = false
-				log.Warning("[PROXY][SD] Service Discovery returned empty host list! Returning last good set.")
-
-				if spec.LastGoodHostList == nil {
-					log.Warning("[PROXY][SD] Last good host list is nil, returning empty set.")
-					spec.LastGoodHostList = apidef.NewHostList()
-				}
+		if err != nil {
+			spec.ServiceRefreshInProgress = false
+			return nil, err
+		}
+		// Set the cached value
+		if data.Len() == 0 {
+			spec.HasRun = true
+			spec.ServiceRefreshInProgress = false
+			log.Warning("[PROXY][SD] Service Discovery returned empty host list! Returning last good set.")
 
-				return spec.LastGoodHostList, nil
+			if spec.LastGoodHostList == nil {
+				log.Warning("[PROXY][SD] Last good host list is nil, returning empty set.")
+				spec.LastGoodHostList = apidef.NewHostList()
 			}
 
-			ServiceCache.Set(spec.APIID, data, cache.DefaultExpiration)
-			// Stash it too
-			spec.LastGoodHostList = data
-			spec.HasRun = true
-			spec.ServiceRefreshInProgress = false
-			return data, err
+			return spec.LastGoodHostList, nil
 		}
+
+		ServiceCache.Set(spec.APIID, data, cache.DefaultExpiration)
+		// Stash it too
+		spec.LastGoodHostList = data
+		spec.HasRun = true
 		spec.ServiceRefreshInProgress = false
-		return nil, err
+		return data, nil
```

Comment:
Note the old `if err == nil` at the top, that is now `if err != nil`. We were using a non-error early return, which was weird. We now use a regular error early return. So this is no longer the "there was an error" branch, this is the "all went ok" branch.

This is perhaps confusing because both have a `spec.ServiceRefreshInProgress = false` before the return.